core/App/ehsm_provider.cpp

```diff
@@ -194,66 +273,94 @@ ehsm_status_t CreateKey(ehsm_keyblob_t *cmk)
     }
 
     /* this api only support for symmetric keys */
     if (cmk->metadata.keyspec != EH_AES_GCM_128 &&
-        cmk->metadata.keyspec != EH_SM4) {
+        cmk->metadata.keyspec != EH_AES_GCM_192 &&
+        cmk->metadata.keyspec != EH_AES_GCM_256 &&
+        cmk->metadata.keyspec != EH_SM4)
+    {
         return EH_KEYSPEC_INVALID;
     }
 
     /* only support to directly encrypt data of less than 6 KB */
     if (plaintext->data == NULL || plaintext->datalen == 0 ||
-        plaintext->datalen > EH_ENCRYPT_MAX_SIZE) {
+        plaintext->datalen > EH_ENCRYPT_MAX_SIZE)
+    {
         return EH_ARGUMENTS_BAD;
     }
 
-    switch(cmk->metadata.keyspec) {
-        case EH_AES_GCM_128:
-            /* calculate the ciphertext length */
-            if (ciphertext->datalen == 0) {
-                ciphertext->datalen = plaintext->datalen + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE;
-                return EH_OK;
-            }
-            /* check if the datalen is valid */
-            if (ciphertext->data == NULL ||
-                ciphertext->datalen != plaintext->datalen + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE)
-                return EH_ARGUMENTS_BAD;
-
-            if ((0 != aad->datalen) && (NULL == aad->data)) {
-                return EH_ARGUMENTS_BAD;
-            }
-
-            if ((0 == aad->datalen) && (NULL != aad->data)) {
-                return EH_ARGUMENTS_BAD;
-            }
-
-            ret = enclave_aes_encrypt(g_enclave_id,
-                                  &sgxStatus,
-                                  aad->data,
-                                  aad->datalen,
-                                  cmk->keyblob,
-                                  cmk->keybloblen,
-                                  plaintext->data,
-                                  plaintext->datalen,
-                                  ciphertext->data,
-                                  ciphertext->datalen);
-            break;
-        case EH_SM4:
-            //TODO
-            break;
-        default:
-            return EH_KEYSPEC_INVALID;
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_AES_GCM_128:
+    case EH_AES_GCM_192:
+    case EH_AES_GCM_256:
+        /* calculate the ciphertext length */
+        if (ciphertext->datalen == 0)
+        {
+            ciphertext->datalen = plaintext->datalen + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE;
+            return EH_OK;
+        }
+        /* check if the datalen is valid */
+        if (ciphertext->data == NULL ||
+            ciphertext->datalen != plaintext->datalen + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE)
+            return EH_ARGUMENTS_BAD;
+
+        if ((0 != aad->datalen) && (NULL == aad->data))
+        {
+            return EH_ARGUMENTS_BAD;
+        }
+        ret = enclave_encrypt(g_enclave_id,
+                              &sgxStatus,
+                              cmk,
+                              SIZE_OF_KEYBLOB_T(cmk->keybloblen),
```

Comment:
this marco seems a little weird

Reply:
I don't quite understand what you mean. Could you describe it in detail